### DISCUSSION TITLE
vndk

### DISCUSSION THREAD

**Discussion by sanketmr121** [phhusson/treble_experimentations#1660] [Q&A]
can someone explain what is vndk and which gsi to download for which phone ( architecture wise).
getting confused by so many new terms in android 11 gsi.

**Comment by kannony** [comment]
@sanketmr121 I believe VNDK and Binder64 are reverse-compatibility modules for phones that were upgraded from Android 8.0 or 8.1. There's [mention of them in the documentation for GSI ROMs](https://source.android.com/setup/build/gsi).

I think most people do not need them.

**Reply by sanketmr121** [reply]
Thanks, that helped.
Just to be sure, I have Samsung Galaxy J6 which was shippped with android 8, so i need to flash a GSI with VNDK module, Right ?

**Reply by kannony** [reply]
I do not know for certain whether that's the right image for the J6. As long as you do a full backup, you limit the risk of having to reinstall everything. VNDK stands for (I think) Vendor Native Development Kit and it has to do with how your manufacturer set up the default ROM on your phone. When they set it up for Android 8.0 or 8.1, it can have dependencies that are required in the image that you install.

The J6 uses a 64-bit processor so it probably doesn't need Binder64, which is for running 32-bit on 64-bit ARMv8 processors.

IIRC, A-B images are for devices upgraded from Android 9 and above. And the A-image is for Android 8. But I'm not sure about that. I do not know whether the A-B partition images are universal but the name suggests that it is.

The best course of action is to make a full backup and try the images in order of how likely they'll work on your system. 

It's super important that you read the documentation. Here's a [quote from the docs](https://source.android.com/setup/build/gsi#vendor-binaries-and-vndk-dependencies):

>The most common supported use case is #2, where the legacy GSIs support devices running >Android 8.1 that were built with BOARD_VNDK_VERSION set to current.

Use case 0 and 1 are not supported by "GSI legacy" according to the documentation. Since the J6 was upgradeable to Android 10, it might be possible to use the "system-roar-arm64-ab-gapps.img.xz" GSI ROM, but I do not know for certain. You absolutely need to do a full backup and be prepared for the possibility that something will go wrong.

**Comment by sanketmr121** [comment]
Thanks for alll this information, cleared many doubts.

**Reply by kannony** [reply]
Oh, one more thing: in the GSI documentation it says A-B images are for legacy (upgraded from 8.0) and that A-images are for Android-10  compatible devices (non legacy).

However, I read the opposite about Phhusson's GSI ROM from someone who had asked the question much earlier. (A-B was for Android 10 compatible devices and A-images were for Android 8 devices). It's very confusing.

I am using an A-B image on an Android-10 compatible device and it works fine though.

**Comment by sanketmr121** [comment]
Tried few gsi and all give this error "failed to mount /system_root invalid argument". Found out it has something to do with dynamic partition, Any help ?

**Reply by kannony** [reply]
Unfortunately, I don't know Samsung phones very well and haven't installed a ROM on one in years. Had you previously installed a custom ROM on the J6? Because if you didn't, this might be the result of an error in preparing the phone for a custom ROM.

**Reply by sanketmr121** [reply]
Been using official Havoc os as my daily driver for 6 months, no issues faced, fantastic performance.
Tried many andoid 9 and 10 gsi before.
Facing this issue for android 11 gsi only.

**Reply by kannony** [reply]
It might be a dirty install, but I can't say for certain. I don't want to give you bad information, but when I've had this error in the past, it was because I didn't wipe system. In your case, it's not clear, unfortunately.

How did you prepare the phone before installing the GSI ROM?

**Reply by sanketmr121** [reply]
boot into recovery
take backup
wipe system, data, vendor, cache, dalvik cache
reboot into recovey
flash system it's - it says image flashed successfully 
then when i reboot into recovey, it says "failed to mount /system_root invalid argument" and reboots in a bootloop

**Reply by kannony** [reply]
Ah, I see, that's what you meant by the dynamic partition causing the problem. That shouldn't be a huge issue. It may be the partition that's the problem.

[There's a YouTube video ](https://www.youtube.com/watch?v=T-7neHzSIvc)which shows a fix that basically runs partition repair, followed by converting to FAT or exFAT, and then converting the partition back to EXT4. Have you seen this one yet? It might work for you.

**Reply by sanketmr121** [reply]
seen it, tried it, same error.

**Reply by kannony** [reply]
What version of TWRP do you have? Have you tried flashing the newest version of TWRP?

EDIT: I don't know the J6 at all but [someone made an unofficial port of TWRP](https://forum.xda-developers.com/t/update-twrp-3-5-0-unofficial-galaxy-j6.4211709/)

**Reply by sanketmr121** [reply]
3.5.0.9.0
the same one in above link

**Reply by kannony** [reply]
I should have noticed that you responded in that thread. :-)

I don't know what the best course of action would be at this point. It may be that the ROM is not compatible with your device. Samsung devices have always been more difficult to work with and it's possible that @phhusson may want to inquire about your particular case, if they have time for it.

One last thing: have you tried reinstalling root?

**Reply by sanketmr121** [reply]
what is reinstalling root, how to do it ?

**Reply by kannony** [reply]
https://www.techradar.com/how-to/how-to-root-your-android-phone-using-windows-10-and-twrp-recovery

I don't think it will help, but it's worth trying.

**Reply by kannony** [reply]
Also IIRC, TWRP has a reinstall root option.

**Reply by sanketmr121** [reply]
Havoc OS android 11 gsi booted succesfuly on first try itself, none additional steps required.

**Reply by rajshekhar99** [reply]
Try vndklite version of gsi, they have r/w access.
